### DISCUSSION TITLE
Getting 137 Error for JS in CodeQL

### DISCUSSION THREAD

**Discussion by PoleMadhu123** [community/community#56934] [Code Security] [answered]
### Select Topic Area

Question

### Body

Here is the Error Logs:

Warning: 01 07:08:21] [build-stdout] [216.418s][warning][gc,alloc] pool-1-thread-1: Retried waiting for GCLocker too often allocating 256 words
  [2023-06-01 07:13:39] [build-stderr] Out of memory while extracting a file.
  Error: 6-01 07:13:42] [ERROR] Spawned process exited abnormally (code 137; tried to run: [/opt/hostedtoolcache/CodeQL/2.13.1-20230428/x64/codeql/javascript/tools/autobuild.sh])
  A fatal error occurred: Exit status 137 from command: [/opt/hostedtoolcache/CodeQL/2.13.1-20230428/x64/codeql/javascript/tools/autobuild.sh]
  Error: A fatal error occurred.
  
  I dont have idea why does this happens. please let me know any solutions.

**Comment by hmakholm** [comment]
Exit status 137 means that the autobuilder process received a SIGKILL (137 = 128+9). The most usual cause for that is that the Linux kernel kills random memory-hungry processes when it cannot satisfy all memory demands from its available RAM and swap.

Given these symptoms here, it could look like perhaps you have an extremely large JS source file that the CodeQL extractor cannot parse within the Action runner's memory?

**Reply by PoleMadhu123** [reply]
Thanks For the Response @hmakholm ,
I added --ram=8000 command in the CodeQL initialization step. so, now it's working Fine and we are getting alerts also.
Have One Doubt, we are using ubuntu-latest runner, as per the documentation linux has 6GB memory. Now I increased runner memory, CodeQL working Fine. Is it correct way to increase the runner memory? give me on explanation on this @hmakholm?

**Reply by hmakholm** [reply]
Hmm, it's pretty curious that setting `--ram` larger than the actual RAM that the runner has would make a 137 problem go away.

Generally, there are two kinds of RAM limits involved: (a) How much memory CodeQL _thinks_ it can use, and (b) how much memory the OS is willing to give it. If the analysis cannot completed within limit (a), you would usually see an error that explicitly complains about being out of memory (the exact wording of this depends on which part of the system runs up against is limit), but the exit-137 problem means that CodeQL itself is happy but the OS kicked it out because (a) was larger than (b).

What the `--ram` option does is to override CodeQL's default assumption about (a). By default it will _try_ to keep it a certain fraction of the physical memory available, but OSes don't make it easy to find out that from a program in a portable and reliable way. So usually we advise setting `--ram` to a value a bit _below_ the physical RAM, say 500 MB less than the actual size.

(For completeness, there are also some components of the system that _cannot_ regulate their RAM usage, so they ignore (a) and just hope they can do what they need to do within (b). They _ignore_ the `--ram` option.)

In any case, if you managed to make the acute problem go away, then you may not want to delve too deeply into that. I would, however, recommend trying if `--ram=5500` is _also_ sufficient to avoid the problem; if it does, that would feel like a more robust long-term solution. I hope this explanation can be useful in case the problem recurs.

**Reply by PoleMadhu123** [reply]
Thanks for the explanation @hmakholm 
I tried --ram=5000, 5500, 6000 getting Out Of Memory errors.
When I'm not mentioned or passing any ram value to initialize step, CodeQL takes this RAM Value "CODEQL_RAM: 5919".
We are confusing now, what happens after increase the RAM size. If it is okay to up the storage limit for getting alerts, then we are good to use CodeQL for JS project.
If it is not correct way to increase the RAM size value for generating alerts, then we can stop using the CodeQL for JS project.
Please suggest your opinions on this.

**Reply by hmakholm** [reply, marked as answer]
Hmm, I'm at a loss as to what exactly is happening here. I notice that the GitHub hosted runners [actually have 7 GB](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) rather than 6 as we assumed, and that might make a difference. But other than that ...

In any case, if you find it works in practice with `--ram=8000` then go for that.
If the analysis completes without explicitly hitting an error, then you should be able to trust that all the alerts that _should_ be generated _have_ also been generated -- we won't silently omit any because of RAM trouble.

**Reply by PoleMadhu123** [reply]
Thanks for the suggestions. @hmakholm 
As you mentioned that GitHub hosted runners have 7GB Memory. Now, First I tried with 7000 it's working and then I reduced to 6500, it's working fine and getting alerts.
As per this Documentation they mentioned 6GB RAM for Linux or ubuntu server. (We are using ubuntu-latest)
https://github.com/github/codeql-action/blob/main/analyze/action.yml, Please check this Documentation Once.
Thanks for your help! I can use --ram=6500 for Now.

**Reply by ninjab3s** [reply]
I am running into the same problem. May I ask how you have changed this parameter?

**Reply by PoleMadhu123** [reply]
@ninjab3s can you share the error logs and workflow (yml) file?

**Reply by ninjab3s** [reply]
Hey PoleMadhu123, thank your for asking but I have already solved it. For everyone else who runs into this as well do the following:

create a config file in  `.github/codeql/codeql-config.yml`
```
name: "CodeQL config"
ram: 6500 
```

Then add this config file like this to your `.github/workflows/codeql.yml`:
```
      uses: github/codeql-action/init@v2
      with:
        languages: ${{ matrix.language }}
        config-file: ./.github/codeql/codeql-config.yml
```

**Reply by PoleMadhu123** [reply]
Awesome !!

**Reply by keithhuster** [reply]
I am running into the same `exit 137` error as the original issue poster...

I attempted to set `ram: 1500` in my CodeQL config file since the current self-hosted runner only has 2GB of memory available. However, the CodeQL executions are still failing with the same error. I also noticed in the log file that there is an `Invalid property specified in the configuration file. Ignoring it and proceeding.` message which was not there in prior runs. So, it appears that the manually specified ram limit is being ignored. This was confirmed in the logs by finding that the `github/codeql-action/autobuild@v2` and `github/codeql-action/analyze@v2` actions are configured to use `CODEQL_RAM: 29548`.

I cannot find any documentation regarding usage of the `ram:` property in the CodeQL config file. Would someone be able to point me to the appropriate doc so that I can further troubleshoot this issue?
